proposals/improved-hash-module.md

```diff
@@ -0,0 +1,134 @@
+        
+        # How to combine a hash of hashes ???
+```
+As you can see above we, computed hashes for all of the struct fields, but we are uncertain how to combine those values in a way which produces a good (non compromised) hash value. Python (docs)[https://docs.python.org/3/reference/datamodel.html#object.__hash__] suggest to pack fileds into a tuple and hash the tuple, but this is not possible at this point in time in Mojo.
+
+# Proposal
+
+In order to improve the hash module and address the flaws of the `Hashable` trait, we need to apply following steps.
+
+## Introduce a `Hasher` trait
+
+By introducing a `Hasher` trait we define an abstraction for the hashing algorithm itself, which allows streaming creation of the hash value. Bellow is a possible API of the `Hashable` trait.
+
+```mojo
+trait Hasher:
+    fn __init__(inout self):
+        ...
+    fn update[T: DType](inout self, value: SIMD[T, 1]):
+        ...
+    fn update(inout self, pointer: DTypePointer[DType.uint8], length: Int):
+        ...
+    fn finish(owned self) -> UInt64:
+        ...
+```
+
+## Implement a default `Hasher` in standard library
+
+The standard library should provide a default `Hasher` implementation, but it would be possible for the developers to implement, or choose other hash algorithms, if they better fit their use case.
+
+Bellow you can see a dummy implementation of a `DefaultHasher`
+
+```mojo
+struct DefaultHasher(Hasher):
+    var hash: UInt64
+
+    fn __init__(inout self):
+        self.hash = 42
+    fn update[T: DType](inout self, value: SIMD[T, 1]):
+        ...
+    fn update(inout self, pointer: DTypePointer[DType.uint8], length: Int):
+        ...
+    fn finish(owned self) -> UInt64:
+        return self.hash
+```
+
+## Redesign the `Hashable` trait to follow the data flow principles
+
+Given the `Hasher` trait, we can define the `Hashable` trait to adopt the data flow paradigm instead of call return.
+
+```mojo
+trait Hashable:
+    fn hash_with[H: Hasher](self, inout hasher: H):
+        ...
+```
+
+## Example for `Hashable` struct implementation
+
+The implementation of `Hashable`, where all the fields are `Hashable` is trivial and could be easily synthesized by the compiler. But even if not, the data flow API guides the developer toward a very simple solution
+
+```mojo
+@value
+struct Person(Hashable):
+    var name: String
+    var age: UInt8
+    var friends_names: List[String]
+
+    fn hash_with[H: Hasher](self, inout hasher: H):
+        # self.name.hash_with(hasher), when String is Hashable, otherwise:
+        hasher.update(self.name._as_ptr().bitcast[DType.uint8](), len(self.name))
+        # self.age.hash_with(hasher), when SIMD is hashable, otherwise
+        hasher.update(self.age)
+        # self.friends_names.hash_with(hasher), when List of Hashable types is Hashable, otherwise:
+        for friend in self.friends_names:
+            hasher.update(friend[]._as_ptr().bitcast[DType.uint8](), len(friend[]))
```

Comment:
I propose an alternative here, as it allows the implementer of "hash_with" to use only one type of call instread of two, thus simplifying the API.
If we modify the `Hasher` trait to include a default implementation:
```mojo
trait Hasher:
    fn update(inout self, value: Hashable):
        # this is a default implementation and should not be changed
        value.hash_with[Self](self)
```
then, if the struct attribute has a `hash_with` method or not, the same method is called and users don't have to remember two different ways of calling the hasher:
```mojo
    fn hash_with[H: Hasher](self, inout hasher: H):
        # hasher.update(hasher), when String is Hashable, otherwise:
        hasher.update(self.name._as_ptr().bitcast[DType.uint8](), len(self.name))
        hasher.update(self.age)
        # hasher.update(hasher), when List of Hashable types is Hashable, otherwise:
        for friend in self.friends_names:
            hasher.update(friend[]._as_ptr().bitcast[DType.uint8](), len(friend[]))
```

Reply:
I simplified the API and provide a working example in https://gist.github.com/mzaks/aa66c831dc5e177c2322d5088aac76aa
Let me know if it resolves your concerns.